drivers/include/tla2528.hpp

```diff
@@ -0,0 +1,135 @@
+#pragma once
+#include <libhal/i2c.hpp>
+#include <libhal/steady_clock.hpp>
+#include <libhal/units.hpp>
+
+namespace sjsu::drivers {
+
+// adapters
+class tla2528_adc;
+class tla2528_input_pin;
+class tla2528_output_pin;
+
+/**
+ * @brief tla2528 gpio expander & adc driver
+ *
+ * tla2528 has 8 pins which can be independently operated as an adc,
+ * digital input, and digital out over i2c. The i2c address is configured by
+ * resistors on the chip There are no options for internal pull up or pull down
+ * resistors. The output pins have the option of push-pull or open-drain. When
+ * in adc mode there is an option to increase reading granularity though
+ * sampling averaging (unimplemented).
+ */
+class tla2528
+{
+public:
+  enum class pin_mode : hal::byte
+  {
+    analog_input,
+    digital_input,
+    digital_output_open_drain,
+    digital_output_push_pull
+  };
+
+  tla2528(hal::i2c& p_i2c, hal::byte p_i2c_address = default_address);
+  /**
+   * @brief set what service a pin will provide
+   */
+  void set_pin_mode(pin_mode p_mode, hal::byte p_channel);
+
+  // If channel is not set to correct config then undefined behavior will occur
+  /**
+   * @brief set output level pin
+   *
+   * If the pin is not an output pin the level will be used once it changes to
+   * an output pin.
+   */
+  void set_digital_out(hal::byte p_channel, bool level);
+  /**
+   * @brief set output levels on pins
+   *
+   * @param p_value The byte is used as a bit feild of bool values. i.e the 0th
+   * bit in the byte will set the 0 pin.
+   *
+   * If the pin is not an output pin the level will be used once it changes to
+   * an output pin output.
```

Comment:
Can you explain how this works a bit more? My guess is that the output state is cached within the device and if the direction is set to output pin, it will have this value. If so, we should explain that. We should also say that this API does not change pin direction and thus, the state of the pin will not change for pins set as input pins.

Reply:
updated documentation, look it over